### DISCUSSION TITLE
Generating Python Library from C and Header Files Using SWIG

### DISCUSSION THREAD

**Discussion by ilyasbsran** [swig/swig#2961] [Q&A] [answered]
Hello everyone,

I have been working on a project where I need to generate a Python library from existing C and header files using SWIG. To automate this process, I wrote a generate.sh script. However, I am encountering an issue when trying to use the generated Python library.

**The Problem:**
When I attempt to import and use the generated Python module, I receive the following error:
`ImportError: dynamic module does not define module export function (PyInit_ccl)`

**Steps Taken:**

1. Wrote generate.sh Script:
The script performs the following steps:

- Uses SWIG to generate the necessary wrapper code.
- Compiles the C code and the SWIG-generated wrapper code into a shared library (_ccl.so).

Here is a simplified version of the generate.sh script:
`#!/bin/bash

# Check the number of arguments
if [ "$#" -ne 1 ]; then
  echo "Incorrect number of arguments"
  echo "Usage: $0 <dbc_file>"
  exit 1
fi

# Check if the build directory exists and remove it
build_dir="build"
if [ -d "$build_dir" ]; then
  if [ -w "$build_dir" ]; then
    rm -rf "$build_dir"
    echo "Build directory removed."
  else
    echo "Error: You do not have permission to remove the build directory."
    exit 1
  fi
fi

# Get the directory argument
dir="$1"
base_dir=$(dirname "$0")
log_dir="${base_dir}/build/log"
log_file="${log_dir}/ccl_script.log"

# Function to log messages
log_message() {
  echo "$(date +'%Y-%m-%d %H:%M:%S') - $1" | tee -a $log_file
}

# Create log directory if it does not exist
mkdir -p "$log_dir"

# Check if log directory was created successfully
if [ -d "$log_dir" ]; then
  log_message "$log_dir directory created successfully or already exists."
else
  echo "Failed to create log directory."
  exit 1
fi

# Check if log file is writable
touch "$log_file"
if [ -w "$log_file" ]; then
  log_message "$log_file is writable."
else
  echo "Failed to create or write to log file."
  exit 1
fi

# Check if the directory/file exists
if [ ! -e "$dir" ]; then
  log_message "Error: $dir not found."
  exit 1
fi

# Check if the file is readable
if [ ! -r "$dir" ]; then
  log_message "Error: You do not have permission to read $dir."
  exit 1
fi

# Extract the filename
filename=$(basename "$dir")

# Check if the file is a .dbc file
if [[ $filename != *.dbc ]]; then
  log_message "This is another type of file. It should be a .dbc file."
  exit 1
fi

log_message "This is a .dbc file"

# Check if we have the necessary permissions to use the dbc file with cantools
if [ ! -r "$dir" ] || [ ! -s "$dir" ]; then
  log_message "Error: The dbc file cannot be read or is empty."
  exit 1
fi

log_message "The dbc file has the necessary read permissions and is not empty."

# Check if Python 3 is installed
if ! command -v python3 &>/dev/null; then
  log_message "Python 3 is not installed"
  exit 1
fi

log_message "Python 3 is installed"

MODULE_NAME="cantools"

# Check if the 'cantools' module is installed
if ! python3 -c "import $MODULE_NAME" &>/dev/null; then
  log_message "$MODULE_NAME is not installed, installing..."
  pip3 install "$MODULE_NAME"
  if [ $? -ne 0 ]; then
    log_message "Failed to install $MODULE_NAME"
    exit 1
  fi
  log_message "$MODULE_NAME has been successfully installed"
else
  log_message "$MODULE_NAME is installed"
fi

log_message "Generating C code..."
# Generate C source code from the dbc file
python3 -m cantools generate_c_source --database-name ccl "$dir"
if [ $? -ne 0 ]; then
  log_message "Failed to generate C code using the $dir file."
  exit 1
fi

# Check if the generated files exist
if [ ! -e "ccl.h" ] || [ ! -e "ccl.c" ]; then
  log_message "Error: Generated files ccl.h and/or ccl.c do not exist."
  exit 1
fi

log_message "Generated files ccl.h and ccl.c exist."

src_dir="${base_dir}/build/src"
bindings_dir="${base_dir}/build/bindings"
python_dir="${base_dir}/build/python"
wrap_dir="${base_dir}/build/wrap"
obj_dir="${base_dir}/build/obj"
lib_dir="${base_dir}/build/lib"

# Function to check directory and permissions
check_dir_and_permissions() {
  local dir=$1
  if [ ! -d "$dir" ]; then
    mkdir -p "$dir"
    log_message "$dir directory created."
  else
    log_message "$dir already exists."
  fi

  if [ ! -w "$dir" ]; then
    log_message "Error: You do not have permission to write to $dir."
    exit 1
  fi
}

# Check src, bindings, python, wrap, obj, and lib directories and permissions
check_dir_and_permissions "$src_dir"
check_dir_and_permissions "$bindings_dir"
check_dir_and_permissions "$python_dir"
check_dir_and_permissions "$wrap_dir"
check_dir_and_permissions "$obj_dir"
check_dir_and_permissions "$lib_dir"

# Move the generated files to the src directory
mv ccl.h "$src_dir"
mv ccl.c "$src_dir"

# Check if files were moved successfully
if [ ! -e "$src_dir/ccl.h" ] || [ ! -e "$src_dir/ccl.c" ]; then
  log_message "Error: Failed to move C libraries to $src_dir."
  exit 1
fi

log_message "C libraries moved to the $src_dir successfully."

# Create and write to the ccl.i file in the bindings directory
echo "%module ccl

%{
#define SWIG_FILE_WITH_INIT
#include \"ccl.h\"
%}

%include \"ccl.h\"
" > "${bindings_dir}/ccl.i"

# Check if the file was created and has the correct content
if [ ! -e "${bindings_dir}/ccl.i" ] || ! grep -q "%module ccl" "${bindings_dir}/ccl.i"; then
  log_message "Error: Failed to create file ccl.i in $bindings_dir or incorrect content."
  exit 1
fi

log_message "File ccl.i created in $bindings_dir with the correct content."

# Check if SWIG is installed
if ! command -v swig &>/dev/null; then
  log_message "SWIG is not installed."
  exit 1
fi

log_message "SWIG is installed."

# Run SWIG command
swig -python -outdir "$python_dir" -o "$wrap_dir/ccl_wrap.c" -I"$src_dir" "${bindings_dir}/ccl.i"
if [ $? -ne 0 ]; then
  log_message "SWIG command failed."
  exit 1
fi

log_message "SWIG command executed successfully."

# Check if ccl_wrap.c and ccl.py were created
if [ -e "${wrap_dir}/ccl_wrap.c" ] && [ -e "${python_dir}/ccl.py" ]; then
  log_message "Files ccl_wrap.c and ccl.py were created successfully."
else
  log_message "Error: Files ccl_wrap.c and/or ccl.py were not created."
  exit 1
fi

# Check if gcc is installed and print its version
if ! command -v gcc &>/dev/null; then
  log_message "gcc is not installed."
  exit 1
else
  gcc_version=$(gcc --version | head -n 1)
  log_message "gcc is installed. Version: $gcc_version"
fi

# Run gcc command to compile ccl.c
gcc -O2 -fPIC -c "${src_dir}/ccl.c" -o "${obj_dir}/ccl.o"
if [ $? -ne 0 ]; then
  log_message "gcc command failed."
  exit 1
fi

# Check if ccl.o was created
if [ ! -e "${obj_dir}/ccl.o" ]; then
  log_message "Error: ccl.o was not created."
  exit 1
fi

log_message "gcc command executed successfully and ccl.o was created in ${obj_dir}."

# Find the default Python version
default_python_version=$(python2 -c "import sys; print('python' + sys.version[:3])")

# Find the default Python include directory
python_include_dir="/usr/include/${default_python_version}"
if [ ! -d "$python_include_dir" ]; then
  log_message "Error: Default Python include directory $python_include_dir not found."
  exit 1
fi

log_message "Default Python include directory found: $python_include_dir"

# Run gcc command to compile ccl_wrap.c with the found Python include directory
# Also include the directory where ccl.h is located
gcc -O2 -fPIC -c "${wrap_dir}/ccl_wrap.c" -I"$python_include_dir" -I"$src_dir" -o "${obj_dir}/ccl_wrap.o"
if [ $? -ne 0 ]; then
  log_message "gcc command failed."
  exit 1
fi

# Check if ccl_wrap.o was created
if [ ! -e "${obj_dir}/ccl_wrap.o" ]; then
  log_message "Error: ccl_wrap.o was not created."
  exit 1
fi

log_message "gcc command executed successfully and ccl_wrap.o was created in ${obj_dir}."

# Run gcc command to create the shared object (.so) file
gcc -shared "${obj_dir}/ccl.o" "${obj_dir}/ccl_wrap.o" -o "${lib_dir}/_ccl.so"
if [ $? -ne 0 ]; then
  log_message "gcc command to create .so file failed."
  exit 1
fi

# Check if _ccl.so was created
if [ ! -e "${lib_dir}/_ccl.so" ]; then
  log_message "Error: _ccl.so was not created."
  exit 1
fi

log_message "gcc command executed successfully and _ccl.so was created in ${lib_dir}."

exit 0
`

2. Directory Structure: 
`my_project/
├── build/
│   ├── lib/
│   │   └── _ccl.so
│   ├── obj/
│   ├── python/
│   │   └── ccl.py
│   ├── src/
│   │   └── ccl.c
│   └── wrap/
│       └── ccl_wrap.c
├── generate.sh
├── ccl.i
└── my_script.py
`

3. Python Script (my_script.py):
`import os
import sys
import ctypes

# Add the directory containing ccl.py to sys.path
script_dir = os.path.dirname(os.path.abspath(__file__))
parent_dir = os.path.dirname(script_dir)
ccl_python_dir = os.path.join(parent_dir, 'build', 'python')
ccl_lib_dir = os.path.join(parent_dir, 'build', 'lib')

sys.path.append(ccl_python_dir)

import ccl  # Import the generated Python module

# Load the shared library
ccl_path = os.path.join(ccl_lib_dir, '_ccl.so')
ccl_lib = ctypes.CDLL(ccl_path)
`
**Error Encountered:**
When running my_script.py, I encounter the following error:

`ImportError: dynamic module does not define module export function (PyInit_ccl)`

**Additional Information:**
- The ccl.py and _ccl.so files are generated as expected.
- The ccl_wrap.c file includes the necessary SWIG-generated code.
- I am using Python 3.10.

**Request for Assistance:**
Could anyone provide insights into what might be causing this issue or suggest possible solutions? Any guidance on how to properly generate and use the Python library from the C and header files using SWIG would be greatly appreciated.

Thank you in advance for your help!

**Comment by ilyasbsran** [comment]
When I created everything in the same directory, there was no problem and it worked. But I want to do it in different ways. How can I make this happen?

**Comment by ojwb** [comment]
This seems to be a usage question rather than a report of a bug in SWIG so converting to a discussion.  You might find the mailing list is a better place to ask though as many more people read posts there than bug reports or discussions here.

**Comment by utkonos** [comment]
I'm encountering almost the same exception. I found this mention of the exception:
https://pythonextensionpatterns.readthedocs.io/en/latest/miscellaneous.html

Are you using macOS?

I am trying to get just a simple HelloWorld working, and hitting this exception on import. This is the directory structure:

```none
README.md
pyproject.toml
src
src/hc
src/hc/hello.c
src/hc/hello.i
```

This is the `pyproject.toml`

```toml
[build-system]
requires = ["setuptools>=75.6"]
build-backend = "setuptools.build_meta"

[project]
name = "helloc"
version = "0.0.1"
authors = [
    { name = "Foo" }
]
description = "Hello world SWIG C"
readme = "README.md"
requires-python = ">=3.13"
classifiers = [
    "Development Status :: 1 - Planning",
    "License :: Public Domain",
    "Natural Language :: English",
    "Operating System :: OS Independent",
    "Programming Language :: Python",
    "Programming Language :: Python :: 3",
    "Programming Language :: Python :: 3 :: Only",
    "Programming Language :: Python :: 3.13"
]

[tool.setuptools]
ext-modules = [
  { name = "hello", sources = ["src/hc/hello.c", "src/hc/hello.i"] }
]
```

here is `hello.c`

```c
#include <stdio.h>

void say_hello() {
    printf("Hello, World!\n");
}
```

here is the SWIG interface file:

```swig
%module hello

%{
void say_hello();
%}

void say_hello();
```

And finally, here is the output for `nm`

```none
% nm -m lib.macosx-14.0-x86_64-cpython-313/hello.cpython-313-darwin.so | grep Init
0000000000001140 (__TEXT,__text) external _PyInit__hello
0000000000004120 (__DATA,__data) non-external _PyInit__hello.SWIG_module
```

I will post this question to the mailing list as well as requested. I hope they're friendly. I have tried to follow the documentation and examples.

**Reply by utkonos** [reply, marked as answer]
The problem that I was encountering is that the module name in `pyproject.toml` was wrong. It needed to be prefixed with `_`.

```toml
[tool.setuptools]
ext-modules = [
  { name = "_hello", sources = ["src/hc/hello.c", "src/hc/hello.i"] }
]
```

More info:
https://stackoverflow.com/questions/79212853/swig-hello-world-importerror-dynamic-module-does-not-define-module-export-func
